File: SHELL/Bankers.py
# Banker's Algorithm

import copy

allocation = [
    [0, 1, 0],
    [2, 0, 0],
    [3, 0, 2],
    [2, 1, 1],
    [0, 0, 2],
]

max_matrix = [
    [7, 5, 3],
    [3, 2, 2],
    [9, 0, 2],
    [2, 2, 2],
    [4, 3, 3],
]

available = [3, 3, 2]

m = len(allocation[0])
n = len(allocation)

request = [1, 0, 2]


class Safety:
    def __init__(self, allocation: list, available: list) -> None:
        self.work = copy.deepcopy(available)
        self.finish = [False for _ in range(n)]
        self.safe_sequence = list()

        self.available = copy.deepcopy(available)
        self.allocation = copy.deepcopy(allocation)

        self.need = [
            [(max_matrix[i][j] - allocation[i][j]) for j in range(m)]
            for i in range(n)
        ]

    def need_work_comparison(self, i: int) -> bool:
        for j in range(m):
            if self.need[i][j] > self.work[j]:
                return False

        return True

    def execute_algorithm(self) -> bool:
        i = 0
        time = 0
        iteration_count = 0

        while True:
            if not self.finish[i] and self.need_work_comparison(i):
                self.work = add_lists(self.work, self.allocation[i])
                self.finish[i] = True
                self.safe_sequence.append(i)

                iteration_count -= 1

            i += 1
            iteration_count += 1

            if i == n and False in self.finish:
                i = 0

            if not False in self.finish:
                break

            if iteration_count >= n:
                print("System is not in safe state")
                break

            time += 1

        is_safe = True

        if False in self.finish:
            is_safe = False

        if is_safe:
            print("System is in safe state")
            print(self.safe_sequence)

        return is_safe


class Request:
    def __init__(
        self, allocation: list, available: list, request: list, process_no: int
    ) -> None:
        self.allocation = copy.deepcopy(allocation)
        self.available = copy.deepcopy(available)
        self.request = copy.deepcopy(request)
        self.process_no = process_no

    def request_available_compare(self) -> bool:
        for j in range(m):
            if self.request[j] > self.available[j]:
                return False

        return True

    def execute_algorithm(self) -> bool:
        if not self.request_available_compare():
            return False

        for j in range(m):
            self.available[j] -= self.request[j]
            self.allocation[self.process_no][j] += self.request[j]

        print("If the Request is granted, ", end="")

        is_safe = Safety(
            allocation=self.allocation, available=self.available
        ).execute_algorithm()

        if is_safe:
            print("Approving request by adding it to allocation table")

            global allocation
            allocation = copy.deepcopy(self.allocation)

            global available
            available = copy.deepcopy(self.available)
        else:
            print("Request cannot be granted")

        return is_safe


def add_lists(input_list_one: list, input_list_two: list) -> list:
    result = list()

    for i in range(len(input_list_one)):
        result.append(input_list_one[i] + input_list_two[i])

    return result


def print_2d_matrix(matrix: list) -> None:
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            print(f"{matrix[i][j]}\t", end="")

        print()


def print_data() -> None:
    print(f"\nAllocation Table:")
    print_2d_matrix(allocation)
    print(f"\nMax Table:")
    print_2d_matrix(max_matrix)
    print(f"\nAvailable Table:")

    for i in range(len(available)):
        print(f"{available[i]}\t", end="")


def main() -> None:
    Safety(allocation=allocation, available=available).execute_algorithm()

    Request(
        allocation=allocation,
        available=available,
        request=request,
        process_no=1,
    ).execute_algorithm()

    print_data()


if __name__ == "__main__":
    main()

File: SHELL/Bankers.cpp
#include <iostream>
#include <cstring>

#define M 3
#define N 5

struct Constants {
   int process_no;
   int available[M];
   int request[M];
   int allocation[N][M];
   int max_matrix[N][M];
};

Constants constants ={
   .process_no = 1,
   .available  = {3, 3, 2},
   .request    = {1, 0, 2},
   .allocation = {{ 0, 1, 0 },
                  { 2, 0, 0 },
                  { 3, 0, 2 },
                  { 2, 1, 1 },
                  { 0, 0, 2 } },
   .max_matrix = {{ 7, 5, 3 },
                  { 3, 2, 2 },
                  { 9, 0, 2 },
                  { 2, 2, 2 },
                  { 4, 3, 3 } },
};

void add_arrays(int*, int*);

class Safety {
private:
   int  work[M];
   bool finish[N];
   int  safe_sequence[N];
   int  available[M];
   int  allocation[N][M];
   int  need[N][M];

public:
   Safety(int allocation[N][M], int available[M]) {
      memcpy(this->work, constants.available, sizeof constants.available);
      memcpy(this->available, constants.available, sizeof constants.available);
      memcpy(this->allocation, constants.allocation,
             sizeof constants.allocation);

      for ( int i = 0; i < N; i++ ) {
         this->finish[i] = false;
      }

      for ( int i = 0; i < N; i++ ) {
         for ( int j = 0; j < M; j++ ) {
            this->need[i][j] =
               constants.max_matrix[i][j] - constants.allocation[i][j];
         }
      }
   }

   bool need_work_comparison(int i) {
      for ( int j = 0; j < M; j++ ) {
         if ( this->need[i][j] > this->work[j] ) {
            return false;
         }
      }

      return true;
   }

   bool is_false_in_finish() {
      for ( int i = 0; i < N; i++ ) {
         if ( this->finish[i] == false ) {
            return true;
         }
      }

      return false;
   }

   void print_safe_sequence() const {
      for ( int i = 0; i < N; i++ ) {
         std::cout << this->safe_sequence[i] << "\t";
      }

      std::cout << std::endl;
   }

   bool execute_algorithm() {
      int  i               = 0;
      int  iteration_count = 0;
      int  safe_index      = 0;
      bool is_safe         = true;

      while ( true ) {
         if ( ! this->finish[i] && this->need_work_comparison(i) ) {
            add_arrays(this->work, this->allocation[i]);
            this->finish[i]                   = true;
            this->safe_sequence[safe_index++] = i;

            iteration_count -= 1;
         }

         i += 1;
         iteration_count += 1;

         if ( i == N && this->is_false_in_finish() ) {
            i = 0;
         }

         if ( ! this->is_false_in_finish() ) {
            break;
         }

         if ( iteration_count >= N ) {
            std::cout << "System is not in safe state" << std::endl;
            break;
         }
      }

      is_safe = true;

      if ( this->is_false_in_finish() ) {
         is_safe = false;
      }

      if ( is_safe ) {
         std::cout << "System is in safe state" << std::endl;
         this->print_safe_sequence();
      }

      return is_safe;
   }
};

class Request {
private:
   int allocation[N][M];
   int available[M];
   int request[M];
   int process_no;

public:
   Request(int allocation[N][M],
           int available[M],
           int request[M],
           int process_no)
      : process_no(process_no) {
      memcpy(this->allocation, constants.allocation,
             sizeof constants.allocation);
      memcpy(this->available, constants.available, sizeof constants.available);
      memcpy(this->request, constants.request, sizeof constants.request);
   }

   bool request_available_compare() {
      for ( int i = 0; i < M; i++ ) {
         if ( this->request[i] > this->available[i] ) {
            return false;
         }
      }

      return true;
   }

   bool execute_algorithm() {
      bool is_safe = true;

      if ( ! this->request_available_compare() ) {
         return false;
      }

      for ( int i = 0; i < M; i++ ) {
         // std::cout << available[i] << " " << request[i];
         this->available[i] -= this->request[i];
         this->allocation[this->process_no][i] += this->request[i];
         // std::cout << available[i] << " " << request[i];
      }

      Safety safety = Safety(this->allocation, this->available);
      is_safe       = safety.execute_algorithm();

      if ( is_safe ) {
         std::cout << "Approving request by adding it to allocation table"
                   << std::endl;

         memcpy(constants.allocation, this->allocation,
                sizeof this->allocation);
         memcpy(constants.available, this->available, sizeof this->available);
      } else {
         std::cout << "Request cannot be granted" << std::endl;
      }

      return is_safe;
   }
};

void
add_arrays(int array_one[M], int array_two[M]) {
   for ( int i = 0; i < M; i++ ) {
      array_one[i] += array_two[i];
   }
}

void
print_2d_matrix(int matrix[N][M]) {
   for ( int i = 0; i < N; i++ ) {
      for ( int j = 0; j < M; j++ ) {
         std::cout << matrix[i][j] << "\t";
      }

      std::cout << std::endl;
   }
}

void
print_data() {
   std::cout << std::endl << "Allocation Table:" << std::endl;
   print_2d_matrix(constants.allocation);
   std::cout << std::endl << "Max Table:" << std::endl;
   print_2d_matrix(constants.max_matrix);
   std::cout << std::endl << "Available Table: " << std::endl;

   for ( int i = 0; i < M; i++ ) {
      std::cout << constants.available[i] << "\t";
   }

   std::cout << std::endl;
}

int
main(void) {
   system("pause");
   Safety safety = Safety(constants.allocation, constants.available);
   safety.execute_algorithm();

   Request request = Request(constants.allocation, constants.available,
                             constants.request, constants.process_no);
   request.execute_algorithm();

   print_data();

   return 0;
}
